### PR TITLE
Wip/profileobject

### PR DESCRIPTION
**Pull request by sulkaharo**
New profile object, which support the old data format and new time of day based values for all contained data structures, including basal. Converted existing profile code to use the profile object. Added basal plugin, which shows current basal rate and other profile information. Added temporary basal adjustment calculation to BWP, if basal profile is present. Fixed a significant bug in how COB called IOB (passed a date to IOB instead of profile). Test for profile code.

### CODE REVIEW

**Comment by sulkaharo** [comment]
TODO: enable the code to load multiple profiles from the database and use the relevant profile based on when it's been taken into use. This can be implemented without a change in the API for the profile code access.

**Comment by sulkaharo** [comment]
FIXING TESTS!

**Comment by jasoncalabrese** [comment]
really cool, agree that the pebble endpoint needs work, it should feed off the same data and not do new mongo queries

File: lib/plugins/basalprofile.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  basal.updateVisualisation = function updateVisualisation (sbx) {
+  
+    if (!hasRequiredInfo(sbx)) {
+      return;
+    }
+    
+    var basalValue = sbx.data.profile.getBasal(sbx.time);
+    
+    var info = [{label: 'Current basal:', value: basalValue + ' IU'}
```
Comment (jasoncalabrese) [inline lib/plugins/basalprofile.js]: IU or U?  we use U everywhere else

File: lib/plugins/boluswizardpreview.js [hunk 2]
```diff
@@ -100,6 +103,19 @@ function init() {
       , {label: 'Expected effect', value: '-' + results.effectDisplay + ' ' + sbx.units}
       , {label: 'Expected outcome', value: results.outcomeDisplay + ' ' + sbx.units}
     ];
+    
+    if (results.tempBasalAdjustment) {
+      if (results.tempBasalAdjustment.thirtymin > 0) {
+        info.push( {label: '30m temp basal', value: results.tempBasalAdjustment.thirtymin + '%'});
```
Comment (jasoncalabrese) [inline lib/plugins/boluswizardpreview.js]: for some pumps like our animas ping, temp basals are set with +/-%, so 50% = -50%, 110% = +10% and can only be set in 10% increments, so I'd need to add some animas specific display.  Similar to the medtronic specific rounding.  Wonder where that should go.  Guess we'll figure it out later.

**Pull request merged**